### ISSUE TITLE
Make vanillajs score fastest on chrome 124 (again)

### ISSUE THREAD

**Issue by krausest** [krausest/js-framework-benchmark#1661]
With chrome 124 the vanillajs and vanillajs-1 implementation don't score fastest.
Currently it's not clear whether actually some change favors other frameworks or it's a bug with chrome 124 traces.
I'd believe chrome 124 results if we could create a new vanillajs implementation that becomes the fastest implementation again.
The first goal would be an implementation that runs fastest in chrome 124 for create 1,000 rows, replace all rows and create 10,000 rows.
It would be really appreciated to get vanillajs PRs. I promise to run them in a timely fashion. Thanks.

**Comment by antonmak1** [comment]
#1633 Here is written about an assumption to increase the speed of work. I think that after the vanilla-2 js version I will need to close this issue.

**Comment by krausest** [comment]
Thanks. I think I applied the suggestions for vanillajs-2:
*  Spaces in the tr template such that createRow can change nodeValue instead of textContent
* Clear table by cloning the tbody
* (not relevant for the benchmarks below): Use nodeValue instead of data when updating a row
![Screenshot 2024-05-03 at 17 47 47](https://github.com/krausest/js-framework-benchmark/assets/3374055/f60627b0-dfd1-43b3-b6ce-1ac268e04e6e)

Ivi is still faster for create 1k rows, it's a tie for replace rows and vanillajs-2 is faster for create 10k rows.

**Comment by krausest** [comment]
Here's vanillajs with appendInline and the other optimizations applied:
![Screenshot 2024-05-03 at 22 25 03](https://github.com/krausest/js-framework-benchmark/assets/3374055/69e32992-6d4a-4898-8fef-ec93c991c1c9)
Good improvement for create 10k rows. Sadly create 1k rows is stuck and ivi is still ahed.

**Comment by mksunny1** [comment]
Sorry, I have had my hands full the whole time. My implementation in Eventiveness uses the StringBuilder pattern from Java to greatly improve the performance for adding rows, especially for the 10K rows. It's like this:

1. Compose an array containing the outerHTML of all the rows.
2. Join the array at once. This is faster than building up the string incrementally.
3. Create a fragment from the joined string
4. Append the fragment to the tbody.

It is hard to imagine another implementation should be faster than this. You need to realise that faster for 10K rows is always better than for 1k rows. It suggests better scalability.

**Comment by mksunny1** [comment]
it is less performant to store the element template in a 'global' string than just inlining it directly into the 'for' loop used to generate the item markups. I reworked my 'framework' to perform similarly to the inline case.

**Comment by mksunny1** [comment]
Also please don't use multiple calls to push when composing the array. Use a generator and destructure into an array with [...myGenerator]. I forgot this detail earlier.

**Comment by localvoid** [comment]
Another difference that comes to my mind is that ivi uses direct calls to DOM methods to avoid megamorphic callsites [1][2] , but vanillajs implementation doesn't have megamorphic callsites, so switching to direct calls should make vanilla slower.

Also, ivi doesn't use any type of event delegation and there are 2 `addEventListener()` calls per row.

1. https://github.com/localvoid/ivi/blob/bd5bbe8c6b39a7be1051c16ea0a07b3df9a178bd/packages/ivi/src/client/core.ts#L41-L80
2. https://github.com/localvoid/ivi/blob/bd5bbe8c6b39a7be1051c16ea0a07b3df9a178bd/packages/ivi/src/client/core.ts#L589-L593

**Comment by krausest** [comment]
Thanks so far.
I think I can show my concerns with chrome 124 in the following two charts:
This is the duration for rendering. Chrome 124 reports about 2.5 msecs faster rendering duration (and this can indeed be validated in the traces). 
<img width="500" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/fd1201ac-647b-487e-b0d5-d88ea579038f">
That wasn't the case with chrome 123:
<img width="500" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/af5674fa-49ef-4790-a3dd-52c7c1d8020a">
Why is chrome 124 rendering for ivi faster than the others? (Or as for my concerns: Why is chrome _reporting_ faster rendering durations for ivi? I'm still not convinced that those are real.)

The difference in script duration isn't what makes ivi fastest:
<img width="500" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/c770b938-580e-4491-bdd6-1eef4a39b270">

**Comment by krausest** [comment]
One example trace for create 1k rows for ivi:
![Screenshot 2024-05-04 at 10 28 13](https://github.com/krausest/js-framework-benchmark/assets/3374055/b9095a27-2926-43c7-a0e5-880e187d44c9)
Total duration: 34.91 msecs
Script duraton: 3.18 msecs
Rendering: 30 msesc (9.71 msecs recalculate style, **16.87 msecs layout**, 3.22 msecs Pre-Paint, 1.11 msecs Paint)

One example trace for create 1k for vanillajs:
![Screenshot 2024-05-04 at 10 28 36](https://github.com/krausest/js-framework-benchmark/assets/3374055/219b197f-99f0-490b-8db3-e8cfa8f881d7)
Total duration: 38.29 msecs
Script duraton: 3.08 msecs
Rendering: 34 msesc (10.03 msecs recalculate style, **19.66 msecs layout**, 3.45 msecs Pre-Paint, 1.04 msecs Paint)

**Comment by localvoid** [comment]
Trying to find any differences in the DOM output. Maybe because `<a>` elements has classes `lbl` and `remove`, but there aren't any css rules for this classes, so it is highly unlikely. https://github.com/krausest/js-framework-benchmark/blob/f3e82b3ed4cb47edfab85d072adc81f274318239/frameworks/keyed/vanillajs-2/src/Main.js#L9

**Comment by krausest** [comment]
@localvoid Good catch, but it's not causing that effect.
This is create 1k rows for ivi **with** the lbl and remove classes (I'll commit that too) and vanillajs **without**:
<img width=400 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/9367415c-829e-464a-b219-5ab382a44401">
Overall a bad run, but ivi still renders 2.5 msecs faster (says chrome 124)

**Comment by localvoid** [comment]
> I'll commit that too

I think that this classes (`lbl`, `remove`) are used in vanilla because of event delegation. Originally, all implementations like React, etc didn't have this classes: https://github.com/krausest/js-framework-benchmark/blob/970b34f7704b6dbd496533fa25314c694cb3d05f/frameworks/keyed/react-classes/src/main.jsx#L57-L68

**Comment by antonmak1** [comment]
@krausest The bubbling process of selectRow and deleteRow can be stopped on `tr` by `stopPropagation`. That is, do not do `stopPropagation` immediately on a `click`, but do it only after reaching the '`parentNode` of the function' on `tr`. I have this implemented in cample, but I don’t know how to do it in vanilla yet. This will speed up the work.

**Comment by antonmak1** [comment]
@krausest I don't know, you can just take `tr` and set the `cancelBubble` property to `true`. But this is not done this way - you need to use `stopPropagation` somehow.

**Comment by krausest** [comment]
@localvoid You're right. I reverted ivi and removed the classes from the vanillajs* implementations.
The result (total duration) still looks like that:
<img width=500 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/2ece2034-d86f-40c9-84b8-64a7d5f461f7">

Now what's interesting is the results for create 1k rows without warmup (i.e. initBenchmark just checking the existence of #run, no create / clear cycle):
<img width="500" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/c08a3c10-1e0c-46f7-9fa4-adc23691517b">
And even more interesting is the rendering duration without warmup:
<img width="500" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/454a2037-7ee5-45d5-aaeb-00d4cc9ccc7a">
A flatline, which really makes sense (here...).

**Comment by mksunny1** [comment]
@krausest I like this last chart. I think you are looking for a more consistent measurement scheme, that can weed out the noise and measure the most important things about how the frameworks build DOMs. Will a vanillajs implementation that scores faster now satisfy this in the long run?

**Comment by mksunny1** [comment]
It could be worth it to make the measurements more fine-grained here.

**Comment by krausest** [comment]
@robbiespeed good catch, indeed ivi stands out that the dom nodes created by initBenchmark are still reported by chrome (in contrast to vanillajs, doohtml and solid). @localvoid is that intentional?

**Comment by robbiespeed** [comment]
May have figured out the reason for the different node deltas. There's a bit of a hint if you look at the graph somewhere during layout for VanillaJS chrome will create 1k new nodes for the `::before` pseudo element of each row for the remove glyph (removing the glyph classes will make the node count a consistent 10 nodes per row). My guess chrome skips creating new nodes for the pseudo elements if there were old ones still around to "recycle".

It would be interesting to see how the benchmark fairs when removing the css that creates the pseudo elements.

**Comment by localvoid** [comment]
@krausest No, I don't have a clue why this is happening. ivi doesn't have any type of DOM or js object recycling.

**Comment by robbiespeed** [comment]
Ran vanillajs against ivi with and without the glyphs. Removing the glyphs didn't help close the perf gap.

**Comment by krausest** [comment]
@robbiespeed Maybe we're really on to something
Seems like forceGC() which called window.gc() a few times didn't clean the dom nodes for ivi.
If we run window.gc() and HeapProfiler.collectGarbage" the node count goes down:
![Screenshot 2024-05-07 at 20 58 40](https://github.com/krausest/js-framework-benchmark/assets/3374055/274deb83-234e-430e-a595-45d4c3d52842)

And even more, the performance is closer to what we'd expect:
<img width=400 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/e328e375-7f29-4ef9-8da9-58c7def5263d">
And performance doesn't change wildly when a 500 msec delay is added:
<img width=400 src="https://github.com/krausest/js-framework-benchmark/assets/3374055/7f097e1d-7299-4ef9-87ff-513e0cb969ab">

Now I'll run the benchmarks above in a rather clean run with the modified forceGC call. I'll report back.

**Comment by krausest** [comment]
With 1.000 msecs delay the graph looks pretty similar, which is really good news!
<img width="400" src="https://github.com/krausest/js-framework-benchmark/assets/3374055/3f4636dd-8515-47d5-b3ea-19c8735ca34d">

**Comment by krausest** [comment]
Conclusion:
* Somehow calling 7x window.gc() didn't clean up the dom nodes for ivi. 
* Running window.gc() and HeapProfiler.collectGarbage in a loop with seven iterations clears the dom nodes for ivi
* With the nodes gc'ed ivi performs similar to chrome 123 and similar to manual tracing
(Which leaves one question: Why is ivi much faster when the nodes are no gc'ed? Is there some kind of caching in chrome 124?)
* This means I'm pretty confident we can get reasonable numbers for chrome 124 for all create actions. (There's might be still the remove row issue which starting with chrome 124 gives frameworks with RAF an advantage, but I'll save that for the next days...)

@paulirish Whenever it comes to GC we need your help. Running windows.gc() in a loop was your recommendation the last time. Is this still what you'd recommend? Any idea why HeapProfiler.collectGarbage collects more nodes for ivi than windows.gc() alone?

**Comment by robbiespeed** [comment]
@krausest try `window.gc({type:'major',execution:'sync',flavor:'last-resort'})`.

This is what the results I get when doing 7 gc with default params (`{type:'major',execution:'sync'}`):
![Screenshot from 2024-05-07 22-21-54](https://github.com/krausest/js-framework-benchmark/assets/1743161/9c9f4b0a-f3bc-4bf1-accd-bd4571212621)

vs `{type:'major',execution:'sync',flavor:'last-resort'}`:
![Screenshot from 2024-05-07 22-19-38](https://github.com/krausest/js-framework-benchmark/assets/1743161/e6721fb4-7f0f-47b9-94c0-cce3c2ff8607)

I'm not entirely sure of the differences of last-resort GC, but it seems to be a more thorough GC pass meant for when the process has reached it's memory limits. I suspect `HeapProfiler.collectGarbage` is doing a last-resort GC internally.

`gc` options are documented in [v8 source](https://github.com/v8/v8/blob/106f228ca86505edb70aa575c99fd2d636819d17/src/extensions/gc-extension.h#L19)

**Comment by krausest** [comment]
@robbiespeed I'll add that for chrome 125. I just finished chrome 124 for keyed results with window.gc and HeapProfiler.collectGarbage

Issue resolved: We made vanillajs fastest again by slowing down ivi 😄 . (Though not entirely true: We also tweaked vanillajs a bit)
